### ISSUE TITLE
Documentation of _CtxMethodMixin: must be tensors?

### ISSUE THREAD

**Issue by danpovey** [pytorch/pytorch#37751]
The documentation for _CtxMethodMixin.save_for_backward seems to imply that it accepts only tensor arguments:

https://github.com/pytorch/pytorch/blob/843c0230f2928aad61a6940688da3a6cd6c4cd57/torch/autograd/function.py#L13

... as does the function name of `saved_tensors`. https://github.com/pytorch/pytorch/blob/843c0230f2928aad61a6940688da3a6cd6c4cd57/torch/autograd/function.py#L386 .   
and this documentation:
https://github.com/pytorch/pytorch/blob/843c0230f2928aad61a6940688da3a6cd6c4cd57/torch/autograd/function.py#L161

But the documentation here:
https://github.com/pytorch/tutorials/blob/f557ee0addc24900929bafd4ceff2fc4c3f47072/beginner_source/examples_autograd/two_layer_net_custom_function.py#L26
indicates that they can be tensors or other types.

I attempted to read the code (which isn't easy as it involves metaclasses) and as far as I can tell they may be tensors or other types.  I think the documentation should be made consistent on this point.

**Comment by ngimel** [comment]
`save_for_backward` is only useful for tensors, because if tensors were saved for backward, there's a check performed to make sure that they were not modified by inplace operation. All other types that you might need in the backward can be directly saved as attributes of `ctx`. 
You are right, the documentation should be made clear, we'll accept a PR fixing that.

**Comment by albanD** [comment]
Indeed, we could make it clearer that `save_for_backward` must be used if you need to save tensors that are input or output of the `forward`.
If you need to save anything else (non-tensor input/output or intermediary results) you can simply save them in the `ctx` as attributes.

**Comment by danpovey** [comment]
OK..  so I'll work on the assumption that for non-tensor types you can just
do ctx.foo = bar and
have it available in the backward pass.
That would be nice to have documented somehow too, e.g. in the 'for
dummies' example.
I expect a lot of people wonder this kind of thing.

**Comment by danpovey** [comment]
I'm asking someone to help with making a PR.

**Comment by albanD** [comment]
I think the main change we want to do is add "How to save data for the backward?" section to the doc on extending autograd [here](https://pytorch.org/docs/stable/notes/extending.html#extending-torch-autograd) (source code for this is [here](https://github.com/pytorch/pytorch/blob/master/docs/source/notes/extending.rst)). That explains that all data can be saved on the ctx **except** Tensors that are either an input or an output which need to be saved with `save_for_backward()` (to avoid memory leak).

Then we can also mention in the doc for `save_for_backward` [here](https://github.com/pytorch/pytorch/blob/843c0230f2928aad61a6940688da3a6cd6c4cd57/torch/autograd/function.py#L13) that you only need to use this for Tensors that are input/output. Everything else can use the ctx.
Then update the doc for `forward` [here](https://github.com/pytorch/pytorch/blob/843c0230f2928aad61a6940688da3a6cd6c4cd57/torch/autograd/function.py#L161) to mention that input/output Tensors are a special case and should use `save_for_backward` instead of just the context.
